sessionization/sessionization.md

```diff
@@ -0,0 +1,597 @@
+
+The following code implements the merge operation. Once again, the full, feature-complete implementation is available on [<img src="github.png" style="height:16px;"/> apache/beam](https://github.com/apache/beam/blob/3a4d57eb8976c5f503b32d478a80b1800490f66f/runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/Aggregators.java#L231-L271).
+
+```java
+public TreeMap<IntervalWindow, AccumT> merge(TreeMap<IntervalWindow, AccumT> buff1, TreeMap<IntervalWindow, AccumT> buff2) {
+  if (buff1.isEmpty()) {
+    return buff2;
+  } else if (buff2.isEmpty()) {
+    return buff1;
+  }
+  // Init new tree map to merge both buffers
+  TreeMap<IntervalWindow, AccumT> res = zero();
+  PeekingIterator<Map.Entry<IntervalWindow, AccumT>> it1 =
+      Iterators.peekingIterator(buff1.entrySet().iterator());
+  PeekingIterator<Map.Entry<IntervalWindow, AccumT>> it2 =
+      Iterators.peekingIterator(buff2.entrySet().iterator());
+
+  AccumT acc = null;
+  IntervalWindow window = null;
+  while (it1.hasNext() || it2.hasNext()) {
+    // pick iterator with the smallest window ahead and forward it
+    Map.Entry<IntervalWindow, AccumT> nextMin =
+        (it1.hasNext() && it2.hasNext())
+            ? it1.peek().getKey().compareTo(it2.peek().getKey()) <= 0 ? it1.next() : it2.next()
+            : it1.hasNext() ? it1.next() : it2.next();
+    if (window != null && window.intersects(nextMin.getKey())) {
+      // extend window and merge accumulators if windows intersect
+      window = window.span(nextMin.getKey());
+      acc = fn.mergeAccumulators(ImmutableList.of(acc, nextMin.getValue()));
+    } else {
+      // store window / accumulator if necessary and continue with next minimum
+      if (window != null && acc != null) {
+        res.put(window, acc);
+      }
+      acc = nextMin.getValue();
+      window = nextMin.getKey();
+    }
+  }
+  if (window != null && acc != null) {
+    res.put(window, acc);
+  }
+  return res;
+}
+```
+
+## Benchmarks
+
+### Nexmark
+
+Apache Beam contains a benchmark suite called Nexmark consisting of multiple "real-world" queries over a three entities model representing an online auction system. Nexmark doesn't necessarily give an accurate overall number as it's running locally and certain aspects such as serialization and shuffling data are simply ignored.
+
+Nevertheless, it will give us a decent estimate how the optimized aggregator performs compared to an aggregator using the canonical window merging of Beam when reducing partitions. In the following we're looking at Nexmark Query 11 (user sessions) for 1 million  and 10 million events.
+
+##### Results
+
+```vega-lite
```

Comment:
![Screenshot 2022-09-27 at 12 16 17](https://user-images.githubusercontent.com/1401430/192500083-63a4387b-cf4f-4d61-a8f6-ff12d22f1430.png)

Reply:
I did not know vega. Thanks for the pointer.

Reply:
Still I don't find these diagrams very clear. Maybe you should explain that it contains several runs with several results and you depict the different results. Or I misunderstood the diagrams ?

Reply:
yes, you're right here